sys/arm64/arm64/copyinout.S

```diff
@@ -91,23 +105,40 @@ END(copyin)
 /*
  * Copies a string from a user to kernel address
  *
- * int copyinstr(const void *udaddr, void *kaddr, size_t len, size_t *done)
+ * int copyinstr(const void * __capability udaddr, void *kaddr, size_t len,
+ *         size_t *done)
  */
 ENTRY(copyinstr)
 	mov	x5, xzr		/* count = 0 */
 	mov	w4, #1		/* If zero return faulure */
 	cbz	x2, 3f		/* If len == 0 then skip loop */
 
+#if __has_feature(capabilities)
+	cvtd	c1, x1			/* Store through a capability */
+	scbnds	c1, c1, x2		/* Set the bounds */
+#endif
+
 	adr	x6, copyio_fault /* Get the handler address */
 	SET_FAULT_HANDLER(x6, x7) /* Set the handler */
 	ENTER_USER_ACCESS(w6, x7)
 
 	ldr	x7, =VM_MAXUSER_ADDRESS
-1:	cmp	x0, x7
+1:
+#if __has_feature(capabilities)
+	gcvalue	x9, c0
+	cmp	x9, x7
+	b.cs	copyio_fault
+	ldrb	w4, [c0]	/* Load from uaddr */
```

Comment:
Do we want to also use ldtrb here? We could do a bx at the start+end for hybrid kernels?

Reply:
(and a `.arch morello+c64` directive)